dolt-demos/README.md

```diff
@@ -11,10 +10,91 @@ Assuming we have an instance of Dolt, in which we are storing some input data, a
 - given a Flow, a and a set of associated runs, succinctly obtain the data output to Dolt by each run
 
 
-## Capabilities
+## Implementation
+The `metaflow.datatools.DoltDT` class is implemented to effectively serve two use cases: acting as a context manager for reading and writing from Dolt in a running flow, and querying the data read and written by a run of a flow.
+
+### Running Metaflow
+The operations supported by `DoltDT` are reading and writing a Pandas `DataFrame` object to and from Dolt, and committing changes. Those operations are supported by three simple methods on `DoltDT`:
+```python
+def write_table(self, table_name: str, df: pd.DataFrame, pks: List[str]):
+    """
+    Writes the contents of the given DataFrame to the specified table. If the table exists it is updated, if it
+    does not it is created.
+    """
+
+def read_table(self, table_name: str) -> pd.DataFrame:
+    """
+    Returns the specified tables as a DataFrame.
+    """
+
+def commit_table_writes(self, allow_empty=True):
+    """
+    Creates a new commit containing all the changes recorded in self.dolt_data.['table_writes'], meaning that the
+    precise data can be reproduced exactly later on by querying self.flow_spec.
+    """
+```
+
+The goal of this API is to integrate the Dolt commit graph with the Metaflow object hierchy, we do this with two objects:
+```python
+class DoltTableRead:
+    def __init__(self, run_id: int, step: str, branch: str, commit: str, table_name: str):
+        self.run_id = run_id
+        self.step = step
+        self.commit = commit
+        self.branch = branch
+        self.table_name = table_name
+
+
+class DoltTableWrite:
+    def __init__(self, run_id: int, step: str, table_name: str):
+        self.run_id = run_id
+        self.step = step
+        self.commit = None
+        self.branch = None
+        self.table_name = table_name
+
+    def set_commit_and_branch(self, commit: str, branch: str):
+        self.commit = commit
+        self.branch = branch
+```
+
+Every time read, write, or commit, are called `DoltTableRead` and `DoltTableWrite` objects record the state of the flow (`run_id`, `step`) and the state of Dolt (`branch`, `commit`, `table_name`), and store that information in the Metaflows data artifact tracking mechanism.
+
+This means that given a Flow definition, possibly recovered from an S3 bucket, and a clone of the Dolt that Flow's runs were executed against, the reads and writes can be uniquely recovered.
```

Comment:
I feel like this should be said earlier. Like the first sentence of the paragraph. This is the most essential point.

Reply:
Good feedback, thanks